### ISSUE TITLE
CSRF: able to log user out

### ISSUE THREAD

**Issue by engerm** [RagtagOpen/nomad#254]
Logout uses a GET request.

Able to CSRF a user to log them out.

```
GET /logout HTTP/1.1
Host: nomad.ragtag.org
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:55.0) Gecko/20100101 Firefox/55.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Referer: https://nomad.ragtag.org/
Cookie: session=.eJwdz01qAzEMBeCrFK-zkGVZtnKZQdYPLYUWZpJVyN1rutf33tOrHHnG9Vnuj_MZt3J8ebkXiIiqmdIzEcQrzslB7lkRp-MEqdKFAlnBm1vTfTVhYDOM1oII0aWbKgyuSdAbz1Vp2gLvOkYFBhqZNQFXxoohaBZTNzMqt2LXmcfj9zt-9h6T0B3RfesxWAR2B7beCWZfActRVzBv97zi_H_iVT6emy5UDtZKbk4gMqXR5I5Lw1uylff7D6JcSX0.DLsO3Q.z8wHsPEtJsgwfJJni4c3h6am49U; remember_token=b2a6e6a1-4dcd-4099-8934-8652baed3f6c|c9a31b4fcda0c21f507909006f6ea7efd406396c3c6e85f1e54f079ff6a03a0f4120ebda52ae67b081715cb2dffd8f7b52f1f819ccee1d9ae71c894df5a3a8b9
DNT: 1
Connection: close
Upgrade-Insecure-Requests: 1
Pragma: no-cache
Cache-Control: no-cache
```

**Comment by simonl2002** [comment]
I don't see a CSRF token with this request.

**Comment by engerm** [comment]
the GET CSRF no longer works, but you can still CSRF using a POST.

There is no csrf token (as @simonl2002 said)

```
POST /logout HTTP/1.1
Host: nomad-staging.ragtag.org
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:56.0) Gecko/20100101 Firefox/56.0
Accept: */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
X-CSRFToken: ImQ5MzBmN2Y5ZjQxYWI3ZDgxYzEzMjk2YTkwNDlhNWY4NDY1NGIyMDEi.DMVruQ.IGuhgS7R_APs6-Zm8PKWXsufrdU
X-Requested-With: XMLHttpRequest
Referer: https://nomad-staging.ragtag.org/
Cookie: intercom-id-j1u8bznm=37a1ba04-b19a-49f4-b676-751f43f95e8e; _ga=GA1.2.834310421.1508021233; _gid=GA1.2.1159648555.1508021233; session=.eJwNjjuKQzEMAO_iOoUt65vLPCTLghBIILtd2Luvpx9mvu16ZLs3QxgZqKGUS1mStAiX9Y3grmoVkmPKGsjTtxjoZKReIBYrlXTRTlVwjE1bO8RUmQkRtSsyHOaYp8FRwOIHkQgL4OTepd3a-vnU9ft-7tf5SZu9pKxw-CnrWMc2dutoTqXIhAF9tL9_gm015w.DMVSNQ.CQ5eX8UQMJp7rkGA-3hMxXhS7nI; session=.eJwdj0tqAzEQBa8StPZC6r98mUGtVpMQSGDGXhnfPSJvX1S9VznyXNdnuT_O57qV4yvKvXSCFk7mxjFNNNiSafa6CMYw6-kaDXU2EhxLOxgKcU3Q7jOMbfIKMxjki5dVcDTFAPdc6eEDsOF2iCeIjj1V9-4gIbVquZV5nXk8fr_Xz-6JjjU1e1Ib22xtbrrL6JX64DQSJofaNve81vl_4lU-nhtFqcyZbcYEItzyISxK6IxUh5T3-w_L_0k0.DMVrgw.aPDxPunY5BEha3T0MMqy8yACC40; remember_token=36055ff1-cdc2-443f-bda6-56743b5340a6|754bfd13121e43ba8fb413e79d9e94932fb2b860bc90e2f9a86e8c56be89e8a14c26d8ff5a266acb8a3f171779e3a8bc9356fa87cec4492c2039c8e7ba113c7d
DNT: 1
Connection: close
Content-Length: 0
```

Reopening bug.

When retesting, please check the GET and POST requests.

**Comment by drabinowitz** [comment]
Sorry am I misunderstanding something, there's an `X-CSRFToken` header in that request. The app is setup to add them for an same domain xhr requests on the page.

**Comment by jayallen** [comment]
@drabinowitz Not for a `GET` request (see [_template.html#L28](https://github.com/RagtagOpen/nomad/blob/master/app/templates/_template.html#L28)) which is what @engerm crafted in the original report (as opposed to the second one which used `POST`):

```
      if (!/^(GET|HEAD|OPTIONS|TRACE)$/i.test(settings.type) && !this.crossDomain) {
        xhr.setRequestHeader("X-CSRFToken", csrf_token);
      }
```

So I'm guessing that `GET` requests to `/logout` should be disallowed, yes?

**Comment by jayallen** [comment]
Errrm, ignore me.  Still getting familiar with the code.  Reading closer, I now see the GET hole in the original report has already been closed and it's apparently POST that is open somehow.

**Comment by dryan** [comment]
Why are we using JavaScript for logging folks out? Let's just put a tiny form in the dropdown with a csrf token embedded in it.

**Comment by iandees** [comment]
Fixed in #606.
